repose-aggregator/components/filters/keystone-v2-filter/src/test/scala/org/openrepose/filters/keystonev2/KeystoneV2AuthorizationTest.scala

```diff
@@ -0,0 +1,178 @@
+      val config = new RolesList().withRole("banana")
+
+      isUserPreAuthed(config, listOfRoles) shouldBe false
+    }
+
+    it("should be false when there is no configuration") {
+      isUserPreAuthed(null, listOfRoles) shouldBe false
+    }
+  }
+
+  describe("shouldAuthorizeTenant") {
+    it("should return false when unconfigured") {
+      shouldAuthorizeTenant(null, false) shouldBe false
+    }
+
+    it("should return false when pre-authorized") {
+      shouldAuthorizeTenant(new ValidateTenantType, true) shouldBe false
+    }
+
+    it("should return true when not pre-authorized") {
+      shouldAuthorizeTenant(new ValidateTenantType, false) shouldBe true
+    }
+  }
+
+  describe("getMatchingTenant") {
+    val token = ValidToken("", "", Seq.empty, None, None, Some("123456"), Seq("456789","foo:789012", "bar:012345"), None, None, Seq.empty, None, None, None)
+    val config = new ValidateTenantType().withStripTokenTenantPrefixes("foo:/baz:")
+
+    it("should return None when a check shouldn't be done") {
+      getMatchingTenant(config, "123456", false, token) shouldBe None
+    }
+
+    it("should return the tenant when it's the default") {
+      getMatchingTenant(config, "123456", true, token) shouldBe Some("123456")
+    }
+
+    it("should return the tenant when it's in the list") {
+      getMatchingTenant(config, "456789", true, token) shouldBe Some("456789")
+    }
+
+    it("should return the tenant when it's prefixed and the default") {
+      val validToken = ValidToken("", "", Seq.empty, None, None, Some("baz:098765"), Seq("456789","foo:789012", "bar:012345"), None, None, Seq.empty, None, None, None)
+      getMatchingTenant(config, "098765", true, validToken) shouldBe Some("baz:098765")
+    }
+
+    it("should return the tenant when it's prefixed and in the list") {
+      getMatchingTenant(config, "789012", true, token) shouldBe Some("foo:789012")
+    }
+
+    it("should return none when the tenant matches but has the wrong prefix") {
+      getMatchingTenant(config, "012345", true, token) shouldBe None
+    }
+
+    it("should return None when the tenant isn't present") {
+      getMatchingTenant(config, "654321", true, token) shouldBe None
+    }
+  }
+
+  describe("authorizeTenant") {
+    it("should succeed when a tenant check is unneeded") {
+      authorizeTenant(false, Some("thing")).success
+    }
+
+    it("should succeed when there was a tenant match") {
+      authorizeTenant(true, Some("thing")).success
+    }
+
+    it("should fail when there wasn't a tenant match") {
+      authorizeTenant(true, None).failed.get shouldBe a [InvalidTenantException]
+    }
+
+    describe("authorizeEndpoints") {
+      val config = new ServiceEndpointType().withName("banana").withRegion("ord").withPublicUrl("http://foo.com")
+
+      it("should succeed if unconfigured") {
+        authorizeEndpoints(null, false, Success(EndpointsData("", Vector.empty))).success
```

Comment:
Somehow I have never known about the `success` and `failure` wrappers. This is extremely useful.

Reply:
it also gives you a nice way of unpacking them for other comparisons. It's just in most of our cases we get a success of unit.